src/Parallel/Local.hpp

```diff
@@ -25,4 +25,17 @@ auto* local(Proxy&& proxy) {
   return proxy.ckLocal();
 }
 
+/// Wrapper for calling Charm++'s `.ckLocalBranch()` on a proxy
+///
+/// The Proxy must be to a Charm++ group chare or nodegroup chare.
+///
+/// The function returns a pointer to the local group/nodegroup chare.
```

Comment:
Could you elaborate on what is confusing/missing from the current documentation? They are wrappers around `.ckLocal()` and `.ckLocalBranch()` which return pointers to their local objects.

From reading the charm docs, I don't know *why* they are different but charm makes it a point to only use `.ckLocal()` on arrays/array-elements and `.ckLocalBranch()` on groups/nodegroups. For that reason, I think it's best to keep them separate.

Reply:
Well if from the user's (our) perspective `ckLocal()` and `ckLocalBranch()` are the same, only named differently for the chare types, then I think it would be nice if we didn't have to care about it. `Parallel::local` can take care of calling the right function depending on the type.

Reply:
As stated in the documentation, `ckLocalBranch` is always guaranteed to return a pointer to an existing object, either the element of the nodegroup on the node (or the element of the group on the PE) on which the call is made.  

`ckLocal` can only be called on a singleton or element of an array, returning either a pointer to the local object if it exists on the local processor and returns `NULL` if the element does not exist or is on a different processor

Therefore technically you should check to see if a ckLocal call didn't return NULL before using the pointer, but in most (if not all) of our use cases we are getting an object from its own proxy

Reply:
In the docs for these functions, it says that it is the responsibility of the user (not this function) to check that the pointer is valid. I think this makes it ok to combine the functions into a single `Parallel::local`

Reply:
Thanks for explaining @kidder. I think we should combine both into a single `Parallel::local` then. It is guaranteed to return a non-null pointer for nodegroups and groups (since there's always a local object), and it may return null for array elements (since they may live on a different core). Sound good?

Reply:
@nilsvu The reason I opted to mirror the Charm++ design when originally putting this together was exactly what @kidder said above: `ckLocal` and `ckLocalBranch` have different behavior because they act on different objects. In most cases, I felt the programmer _should_ be aware of what chare is being used in a piece of code, and how that impacts the validity of the obtained local pointer. I felt that maintaining two distinct names for two (similar but) different functions was reasonable. Anyway, that was my thinking at the time, but feel free to change however you like 🙂

Reply:
> In most cases, I felt the programmer should be aware of what chare is being used in a piece of code, and how that impacts the validity of the obtained local pointer.

Definitely 👍

> I felt that maintaining two distinct names for two (similar but) different functions was reasonable.

Here I disagree. The two names have no difference in meaning; they might as well be called `ckLocal()` and `ckLocal2()` :D The programmer should remember which share they are working with, and hence, if they need to check the result for `nullptr`, but they should not have to remember which of the two function names applies to the chare. Even if the names are different they have to remember the pointer validity, so different function names don't help there.

Either way, I don't feel strongly enough about this to hold up the PR. If @kidder agrees I suggest @knelli2 you merge `Parallel::local` and `Parallel::local_branch`, and if @kidder disagrees then don't.

Reply:
@kidder ping

Reply:
I would like to suggest that `Parallel::local` be renamed to `Parallel::if_local_get_element` and `Parallel::local_branch` be renamed to `Parallel::get_local_element_of_group`

Reply:
@kidder I see your wish to choose more precise names, but I think we shouldn't put "element" in the names, since that can be confused with our notion of domain elements. Also, one of the functions works with a singleton, which is just a single entity. My proposal is: in this case of controversy let's go with minimal changes here and keep the Charm++ names `local` and `local_branch`, or if you agree with my comment above @kidder we merge them under the name of `local`.

Reply:
I think we should match the Charm++ names, at the very least to not diverge in naming from them too much. If we start renaming everything then we can't defer to their documentation as much, and I think deferring to the Charm++ documentation makes our life easier overall. I think this is the argument I made last time this came up for discussion and why Francois went with this distinction.